### PR TITLE
Redesign function management abstraction

### PR DESCRIPTION
**Pull request by dain**
The redesigns the function APIs in preparation for extending runtime function resolution to connectors.

* Add `FuctionHandle` abstraction which is similar to `TableHandle`.
* Add `FunctionMetadata` abstraction which describes a function represented by a FunctionHandle.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/scalar/MapSubscriptOperator.java [hunk 1]
```diff
@@ -39,18 +40,17 @@
...
 import static io.prestosql.spi.type.TypeSignature.parseTypeSignature;
 import static io.prestosql.spi.type.TypeUtils.readNativeValue;
-import static io.prestosql.sql.relational.Signatures.castSignature;
 import static io.prestosql.util.Reflection.methodHandle;
 import static java.lang.String.format;
 
 public class MapSubscriptOperator
         extends SqlOperator
 {
-    private static final MethodHandle METHOD_HANDLE_BOOLEAN = methodHandle(MapSubscriptOperator.class, "subscript", boolean.class, InterpretedFunctionInvoker.class, Type.class, Type.class, ConnectorSession.class, Block.class, boolean.class);
```
Comment (rongrong) [inline presto-main/src/main/java/io/prestosql/operator/scalar/MapSubscriptOperator.java]: Change in this class should probably be a separate commit in itself.

File: presto-main/src/main/java/io/prestosql/sql/InterpretedFunctionInvoker.java [hunk 2]
```diff
@@ -50,23 +52,45 @@ public Object invoke(Signature function, ConnectorSession session, Object... arg
...
+     * <p>
+     * Returns a value in the native container type corresponding to the declared SQL return type
+     */
+    public Object invoke(FunctionHandle function, ConnectorSession session, List<Object> arguments)
+    {
+        ScalarFunctionImplementation implementation = functionManager.getScalarFunctionImplementation(function);
+        return invoke(implementation, session, arguments);
+    }
+
+    private Object invoke(ScalarFunctionImplementation function, ConnectorSession session, List<Object> arguments)
```
Comment (rongrong) [inline presto-main/src/main/java/io/prestosql/sql/InterpretedFunctionInvoker.java]: Seems that `implementation` is the right name.

File: presto-main/src/main/java/io/prestosql/metadata/FunctionManager.java [hunk 3]
```diff
@@ -60,6 +60,11 @@ public void addFunctions(List<? extends SqlFunction> functions)
         return functionRegistry.list();
     }
 
+    public FunctionHandle lookupFunction(QualifiedName name, List<TypeSignatureProvider> parameterTypes)
```
Comment (rongrong) [inline presto-main/src/main/java/io/prestosql/metadata/FunctionManager.java]: Why is this one not called `resolveFunction`?

File: presto-main/src/main/java/io/prestosql/metadata/FunctionManager.java [hunk 4]
```diff
@@ -109,4 +109,9 @@ public FunctionHandle lookupInternalCastFunction(String name, TypeSignature from
     {
         return functionRegistry.lookupInternalCastFunction(name, fromType, toType);
     }
+
+    public FunctionMetadata getFunctionMetadata(FunctionHandle functionHandle)
```
Comment (rongrong) [inline presto-main/src/main/java/io/prestosql/metadata/FunctionManager.java]: Why this is `getFunctionMatadata(FunctionHandle)` rather than `FunctionHandle.getFunctionMetadata`? Also, conceptually, it seems to me that you'd only need `FunctionMetadata` at plan time, so in some ways `FunctionMetadata` can be `FunctionHandle` (or the other way around). Why do we need both?

File: presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java [hunk 5]
```diff
@@ -264,24 +263,17 @@ protected RowExpression visitBinaryLiteral(BinaryLiteral node, Void context)
         @Override
         protected RowExpression visitGenericLiteral(GenericLiteral node, Void context)
         {
-            Type type;
```
Comment (rongrong) [inline presto-main/src/main/java/io/prestosql/sql/relational/SqlToRowExpressionTranslator.java]: Why is this no longer needed?

File: presto-main/src/test/java/io/prestosql/metadata/TestFunctionRegistry.java [hunk 6]
```diff
@@ -353,16 +352,16 @@ public ResolveFunctionAssertion forParameters(String... parameters)
 
         public ResolveFunctionAssertion returns(SignatureBuilder functionSignature)
         {
-            Signature expectedSignature = functionSignature.name(TEST_FUNCTION_NAME).build();
-            Signature actualSignature = resolveSignature();
-            assertEquals(actualSignature, expectedSignature);
+            FunctionHandle expectedSignature = new FunctionHandle(functionSignature.name(TEST_FUNCTION_NAME).build());
```
Comment (rongrong) [inline presto-main/src/test/java/io/prestosql/metadata/TestFunctionRegistry.java]: `expectedFunction`

File: presto-main/src/test/java/io/prestosql/operator/aggregation/minmaxby/TestMinMaxByAggregation.java [hunk 7]
```diff
@@ -74,23 +79,21 @@ public void testAllRegistered()
...
         assertAggregation(
                 unknownKey,
                 null,
                 createBooleansBlock(null, null),
                 createDoublesBlock(1.0, 2.0));
-        InternalAggregationFunction unknownValue = METADATA.getFunctionManager().getAggregateFunctionImplementation(
-                new Signature("min_by", AGGREGATE, parseTypeSignature(StandardTypes.DOUBLE), parseTypeSignature(StandardTypes.DOUBLE), parseTypeSignature(UnknownType.NAME)));
+        InternalAggregationFunction unknownValue = getInternalAggregationFunction("min_by", DOUBLE, UNKNOWN);
         assertAggregation(
                 unknownKey,
```
Comment (rongrong) [inline presto-main/src/test/java/io/prestosql/operator/aggregation/minmaxby/TestMinMaxByAggregation.java]: I'm guessing this is meant to be `unknownValue`, otherwise `unknownValue` is an unused variable.

File: presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java [hunk 8]
```diff
@@ -926,8 +927,8 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
                 argumentValues.add(value);
                 argumentTypes.add(type);
             }
-            Signature functionSignature = metadata.getFunctionManager().resolveFunction(node.getName(), fromTypes(argumentTypes));
-            ScalarFunctionImplementation function = metadata.getFunctionManager().getScalarFunctionImplementation(functionSignature);
+            FunctionHandle functionHandle = metadata.getFunctionManager().resolveFunction(session, node.getName(), fromTypes(argumentTypes));
+            ScalarFunctionImplementation function = metadata.getFunctionManager().getScalarFunctionImplementation(functionHandle);
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-main/src/main/java/io/prestosql/sql/planner/ExpressionInterpreter.java]: Resolving a function requires a full session, where as everything else in this code only needs a connector session.  The renames are because connector session was called `session`, and I wanted to use that name for the full session.

File: presto-main/src/main/java/io/prestosql/sql/relational/RowExpressionVisitor.java [hunk 9]
```diff
@@ -17,6 +17,8 @@
 {
     R visitCall(CallExpression call, C context);
 
+    R visitSpecialForm(SpecialForm specialForm, C context);
```
Comment (rongrong) [inline presto-main/src/main/java/io/prestosql/sql/relational/RowExpressionVisitor.java]: Seems like it would be more consistent to call this new class `SpecialFormExpression`.
